src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs

```diff
@@ -40,6 +40,19 @@ public sealed partial class TextLoader : IDataLoader
         /// </example>
         public sealed class Column
         {
+            public Column() { }
+
+            public Column(string name, DataKind? type, int index)
+               : this(name, type, new[] { new Range(index) }) { }
+
+            public Column(string name, DataKind? type = null, Range[] source = null, KeyRange keyRange = null)
+            {
```

Comment:
>{ [](start = 12, length = 1)

`Contracts.CheckValue(name, nameof(name)); Contracts.CheckValueOrNull(everything_else);`

Also, what's up with range being nullable?

Reply:
Which range?  `source` or `keyRange`?

Both of them can be null today.

---
In reply to: [211433252](https://github.com/dotnet/machinelearning/pull/698#discussion_r211433252) [](ancestors = 211433252)

Reply:
Yes, that's what I meant. What does `source=null` signify? 

---
In reply to: [211438122](https://github.com/dotnet/machinelearning/pull/698#discussion_r211438122) [](ancestors = 211438122,211433252)

Reply:
It was supposed to mean, "I haven't set it yet".  Just like if you say:
```
new TextLoader.Column() { Name = "foo" }
```
today.

At first, I meant these constructors to be as non-prescriptive as the current parameterless constructors are. And the validation only happens once you actually use the object. But since we are adding Contracts checks into them, that reasoning probably doesn't hold anymore.

I'll remove the default `null` value for `source`, and will also need to remove it for `type`.

Reply:
Oh, I see. Well, I like it better this way. We didn't have 'dual-initialized' objects before in our codebase (either the ctor was completely empty, or it would always construct a valid immutable object), so there's no standard practice to follow in this case.

For the self-documentation perspective, also add `Contracts.CheckValueOrNull(type)`

---
In reply to: [211655552](https://github.com/dotnet/machinelearning/pull/698#discussion_r211655552) [](ancestors = 211655552)